File: paleo_workbench/ui/pages/seismic_control_panel.py
from __future__ import annotations

from PySide6.QtCore import Signal
from PySide6.QtWidgets import QComboBox, QFrame, QLabel, QPushButton, QVBoxLayout

from paleo_workbench.ui import tokens
from paleo_workbench.viz.prediction_helpers import field_value
from paleo_workbench.workflow.seismic_prediction import SEISMIC_DISPLAY_MODES


class SeismicControlPanel(QFrame):
    """Right-hand intelligent analysis summary for the current seismic view."""

    send_requested = Signal()
    display_mode_changed = Signal(str)
    well_tie_toggled = Signal(bool)

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setObjectName("SeismicControlPanel")
        self.setFixedWidth(240)
        self._suppress = False

        layout = QVBoxLayout(self)
        layout.setContentsMargins(
            tokens.PANEL_PADDING,
            tokens.PANEL_PADDING,
            tokens.PANEL_PADDING,
            tokens.PANEL_PADDING,
        )
        layout.setSpacing(tokens.SPACE_2)

        self.title_label = QLabel("智能分析结果")
        self.title_label.setObjectName("MapDockTitle")
        layout.addWidget(self.title_label)

        self.status_value = self._add_value(layout, "分析状态", "待开始")
        self.shape_value = self._add_value(layout, "体数据维度", "—")
        self.horizon_value = self._add_value(layout, "目标层位", "—")
        self.mock_value = self._add_value(layout, "输出性质", "—")
        self.attribute_value = self._add_value(layout, "当前属性", "振幅")

        mode_label = QLabel("显示模式")
        mode_label.setObjectName("WorkFieldLabel")
        layout.addWidget(mode_label)
        self.mode_combo = QComboBox()
        self.mode_combo.addItems(list(SEISMIC_DISPLAY_MODES))
        self.mode_combo.currentTextChanged.connect(self._on_mode)
        layout.addWidget(self.mode_combo)
        # Backward-compatible label mirror
        self.mode_value = self.mode_combo

        self.well_tie_btn = QPushButton("井震标定 (Auto-Tie)")
        self.well_tie_btn.setObjectName("SecondaryButton")
        self.well_tie_btn.setCheckable(True)
        self.well_tie_btn.toggled.connect(self._on_well_tie)
        layout.addWidget(self.well_tie_btn)

        layout.addStretch()
        self.send_btn = QPushButton("发送编图")
        self.send_btn.setObjectName("PrimaryButton")
        self.send_btn.clicked.connect(self.send_requested.emit)
        layout.addWidget(self.send_btn)

    def _add_value(self, layout: QVBoxLayout, label_text: str, value_text: str) -> QLabel:
        label = QLabel(label_text)
        label.setObjectName("WorkFieldLabel")
        layout.addWidget(label)
        value = QLabel(value_text)
        value.setObjectName("WorkFieldValue")
        layout.addWidget(value)
        return value

    def _on_mode(self, text: str) -> None:
        if not self._suppress:
            self.display_mode_changed.emit(text)

    def _on_well_tie(self, checked: bool) -> None:
        if not self._suppress:
            self.well_tie_toggled.emit(bool(checked))

    def set_controls_enabled(self, enabled: bool) -> None:
        self.mode_combo.setEnabled(enabled)
        self.well_tie_btn.setEnabled(enabled)
        self.send_btn.setEnabled(enabled)

    def set_attribute_label(self, label: str) -> None:
        self.attribute_value.setText(str(label or "振幅"))

    def update_state(self, task, volume_shape: tuple[int, int, int] | None = None) -> None:
        self.shape_value.setText(
            " × ".join(str(value) for value in volume_shape) if volume_shape else "—"
        )

        summary = field_value(task, "result_summary", {}) or {}
        meta = field_value(task, "model_metadata", {}) or {}
        horizon = ""
        if isinstance(meta, dict):
            horizon = str(meta.get("target_horizon") or "")
        if not horizon and isinstance(summary, dict):
            horizon = str(summary.get("target_horizon") or "")
        self.horizon_value.setText(horizon or "—")
        self.status_value.setText(field_value(task, "status", "") or "待开始")

        if task is None:
            self.mock_value.setText("—")
            self.set_controls_enabled(False)
            return

        # Honest output labeling (P2): random/mock output must never display
        # as 真实, and heuristic output is not a scientific prediction.
        # Mirrors prediction_evidence_panel.py (review finding I3).
        if summary.get("is_mock"):
            nature = "Mock"
        elif not summary.get("final_scientific_prediction", False):
            nature = "启发式"
        else:
            nature = "科学预测"
        if summary.get("demo") or summary.get("source") == "synthetic/demo":
            nature = f"Demo · {nature}"
        replaceable = "可替换" if summary.get("is_replaceable", False) else "固定"
        self.mock_value.setText(f"{nature} · {replaceable}")
        self.set_controls_enabled(True)

File: libs/ui_wellseis/src/qt/seismic_control_panel.cpp
#include <pwb/ui_wellseis/qt/seismic_control_panel.hpp>

#include <QComboBox>
#include <QLabel>
#include <QPushButton>
#include <QVBoxLayout>

#include <pwb/ui_wellseis/output_labels.hpp>
#include <pwb/ui_wellseis/page_state.hpp>
#include <pwb/ui_wellseis/seismic_attributes.hpp>

namespace pwb::ui_wellseis::qt {

namespace {

QString qs(const std::string& text) {
    return QString::fromUtf8(text.data(), static_cast<int>(text.size()));
}

QLabel* add_value(QVBoxLayout* layout, const QString& label_text,
                  const QString& value_text, QWidget* parent) {
    auto* label = new QLabel(label_text, parent);
    label->setObjectName(QStringLiteral("WorkFieldLabel"));
    layout->addWidget(label);
    auto* value = new QLabel(value_text, parent);
    value->setObjectName(QStringLiteral("WorkFieldValue"));
    layout->addWidget(value);
    return value;
}

}  // namespace

SeismicControlPanel::SeismicControlPanel(QWidget* parent)
    : QFrame(parent) {
    setObjectName(QStringLiteral("SeismicControlPanel"));
    setMinimumWidth(200);

    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(12, 12, 12, 12);
    layout->setSpacing(4);

    auto* title = new QLabel(QStringLiteral("智能分析结果"), this);
    title->setObjectName(QStringLiteral("MapDockTitle"));
    layout->addWidget(title);

    status_value_ = add_value(layout, QStringLiteral("分析状态"),
                              QStringLiteral("待开始"), this);
    shape_value_ = add_value(layout, QStringLiteral("体数据维度"),
                           QStringLiteral("—"), this);
    horizon_value_ = add_value(layout, QStringLiteral("目标层位"),
                               QStringLiteral("—"), this);
    mock_value_ = add_value(layout, QStringLiteral("输出性质"),
                            QStringLiteral("—"), this);
    attribute_value_ = add_value(layout, QStringLiteral("当前属性"),
                                 QStringLiteral("振幅"), this);

    auto* mode_label = new QLabel(QStringLiteral("显示模式"), this);
    mode_label->setObjectName(QStringLiteral("WorkFieldLabel"));
    layout->addWidget(mode_label);
    mode_combo_ = new QComboBox(this);
    for (const std::string& mode : seismic_display_modes()) {
        mode_combo_->addItem(qs(mode));
    }
    connect(mode_combo_, &QComboBox::currentTextChanged, this,
            [this](const QString& text) {
                if (!suppress_) {
                    emit display_mode_changed(text);
                }
            });
    layout->addWidget(mode_combo_);

    well_tie_btn_ =
        new QPushButton(QStringLiteral("井震标定 (Auto-Tie)"), this);
    well_tie_btn_->setObjectName(QStringLiteral("SecondaryButton"));
    well_tie_btn_->setCheckable(true);
    connect(well_tie_btn_, &QPushButton::toggled, this, [this](bool checked) {
        if (!suppress_) {
            emit well_tie_toggled(checked);
        }
    });
    layout->addWidget(well_tie_btn_);

    layout->addStretch();
    send_btn_ = new QPushButton(QStringLiteral("发送编图"), this);
    send_btn_->setObjectName(QStringLiteral("PrimaryButton"));
    connect(send_btn_, &QPushButton::clicked, this,
            &SeismicControlPanel::send_requested);
    layout->addWidget(send_btn_);

    set_controls_enabled(false);
}

void SeismicControlPanel::update_state(
    const PredictionTaskSlice* task,
    std::optional<std::array<std::int64_t, 3>> volume_shape) {
    shape_value_->setText(qs(volume_shape_text(volume_shape)));
    const std::string horizon =
        task != nullptr
            ? target_horizon_of(task->model_metadata, task->result_summary)
            : "";
    horizon_value_->setText(horizon.empty() ? QStringLiteral("—")
                                            : qs(horizon));
    status_value_->setText(
        task != nullptr && !task->status.empty() ? qs(task->status)
                                                 : QStringLiteral("待开始"));
    if (task == nullptr) {
        mock_value_->setText(QStringLiteral("—"));
        set_controls_enabled(seismic_controls_enabled(false));
        return;
    }
    mock_value_->setText(qs(seismic_output_nature(task->result_summary)));
    set_controls_enabled(seismic_controls_enabled(true));
}

void SeismicControlPanel::set_attribute_label(const QString& label) {
    attribute_value_->setText(
        qs(attribute_label_or_default(label.toStdString())));
}

void SeismicControlPanel::set_controls_enabled(bool enabled) {
    mode_combo_->setEnabled(enabled);
    well_tie_btn_->setEnabled(enabled);
    send_btn_->setEnabled(enabled);
}

void SeismicControlPanel::set_display_mode(const QString& mode) {
    suppress_ = true;
    mode_combo_->setCurrentText(mode);
    suppress_ = false;
}

void SeismicControlPanel::set_well_tie_checked(bool checked) {
    suppress_ = true;
    well_tie_btn_->setChecked(checked);
    suppress_ = false;
}

QString SeismicControlPanel::display_mode() const {
    return mode_combo_->currentText();
}

}  // namespace pwb::ui_wellseis::qt
